### PR TITLE
Basic RTP demuxing

### PR DESCRIPTION
**Pull request by LVala**

### CODE REVIEW

File: lib/ex_webrtc/peer_connection.ex [hunk 1]
```diff
@@ -372,6 +380,12 @@ defmodule ExWebRTC.PeerConnection do
     new_transceivers = update_local_transceivers(type, state.transceivers, sdp)
     state = set_description(:local, type, sdp, state)
 
+    demuxer = %{
+      state.demuxer
+      | extensions: SDPUtils.get_extensions(sdp),
+        pt_to_mid: SDPUtils.get_payload_types(sdp)
```
Comment (mickel8) [inline lib/ex_webrtc/peer_connection.ex:386]: I wonder about adding ssrc_to_mid too. Should we then move this to apply_remote_description?

File: test/peer_connection/demuxer_test.exs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  @ssrc 333_333
+  @raw_packet %Packet{
+    payload_type: @payload_type,
+    sequence_number: @sequence_number,
+    timestamp: 0,
+    ssrc: @ssrc,
+    payload: <<>>
+  }
+
+  @packet Packet.encode(@raw_packet)
```
Comment (mickel8) [inline test/peer_connection/demuxer_test.exs:21]: I belive packet and raw_packet should switch with their naming

File: test/peer_connection/demuxer_test.exs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  @packet_mid Packet.encode(
+                Packet.set_extension(@raw_packet, :two_byte, [%Extension{id: 15, data: @mid}])
+              )
```
Comment (mickel8) [inline test/peer_connection/demuxer_test.exs:24]: Better pipe this

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: lib/ex_webrtc/peer_connection.ex (hunk 1 above)
Comment (LVala) [reply lib/ex_webrtc/peer_connection.ex:386]: I'm going to leave it for now, will do some improvements when moving the demuxing to standalone DTLS process.

File: lib/ex_webrtc/peer_connection/demuxer.ex [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  def process_data(demuxer, data) do
+    with {:ok, %Packet{} = packet} <- decode(data),
+         {:ok, demuxer, mid} <- match_to_mid(demuxer, packet) do
+      {:ok, demuxer, mid, packet}
+    end
+  end
+
+  # RFC 8843, 9.2
+  defp match_to_mid(demuxer, %Packet{ssrc: ssrc} = packet) do
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/ex_webrtc/peer_connection/demuxer.ex:32]: ?

File: lib/ex_webrtc/sdp_utils.ex [hunk 5]
```diff
@@ -113,6 +114,7 @@ defmodule ExWebRTC.SDPUtils do
     |> Map.new()
   end
 
+  @spec get_payload_types(ExSDP.t()) :: %{non_neg_integer() => binary()}
```
Comment (mickel8) [inline lib/ex_webrtc/sdp_utils.ex:117]: That's a little bit too generic. You still have to read the code to know what this function returns.

```suggestion
  @spec get_payload_types(ExSDP.t()) :: %{pt :: non_neg_integer() => mid :: binary()}
```

File: lib/ex_webrtc/peer_connection/demuxer.ex [hunk 6]
```diff
@@ -1,13 +1,27 @@
 defmodule ExWebRTC.PeerConnection.Demuxer do
   @moduledoc false
+  # RTP demuxing flow:
+  # 1. if packet has sdes mid extension, add it to mapping ssrc => mid
+  #   - if mapping already exista, raise (temporary)
```
Comment (mickel8) [inline lib/ex_webrtc/peer_connection/demuxer.ex:5]:
```suggestion
  #   - if mapping already exists but is different (e.g. ssrc is mapped to different mid), raise (temporary)
```

File: lib/ex_webrtc/peer_connection/demuxer.ex [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  alias ExRTP.Packet
+  alias ExRTP.Packet.Extension
+  alias ExRTP.Packet.Extension.SourceDescription
+
+  @type t() :: %__MODULE__{
+          ssrc_to_mid: %{non_neg_integer() => {non_neg_integer(), binary()}},
+          extensions: %{non_neg_integer() => {module(), atom()}},
+          pt_to_mid: %{non_neg_integer() => binary()}
+        }
```
Comment (mickel8) [inline lib/ex_webrtc/peer_connection/demuxer.ex:19]: That's too generic

File: lib/ex_webrtc/peer_connection/demuxer.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (mickel8) [reply lib/ex_webrtc/peer_connection/demuxer.ex:32]: I mean that we should check whether we don't get mid that was not in the offer/answer

File: lib/ex_webrtc/peer_connection/demuxer.ex [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+          _other ->
+            nil
+        end
+      end)
+
+    case Map.get(demuxer.ssrc_to_mid, ssrc) do
+      {last_mid, last_sn} when mid != nil and mid != last_mid and sn > last_sn ->
+        # temporary, as we belive this case shouldn't occur
+        raise "Received new MID for already mapped SSRC"
```
Comment (mickel8) [inline lib/ex_webrtc/peer_connection/demuxer.ex:58]:
```suggestion
      {last_mid, last_sn} when mid != nil and mid != last_mid ->
        # temporary, as we believe this case shouldn't occur
        raise "Received new MID for already mapped SSRC. SSRC table: #{inspect(demuxer.ssrtc_to_mid}, packet ssrc: #{ssrc}, packet  mid: #{mid}"
```

File: lib/ex_webrtc/peer_connection/demuxer.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (LVala) [reply lib/ex_webrtc/peer_connection/demuxer.ex:32]: Will take care of that in the next pr.

**Review by mickel8 — APPROVED** [review]

**Pull request merged**